### DISCUSSION TITLE
Removing kube-proxy and kube-flannel after cluster created

### DISCUSSION THREAD

**Discussion by jlowe64** [siderolabs/talos#8037] [Q&A] [answered]
Hello,

Is there any way to remove kube-proxy and flannel after the cluster has been created in order to use fully the Cilium CNI? I've tried to do this multiple times but any time the system reboots or I perform a k8s upgrade, even if the machine config is changed, it just resorts back to kube-proxy and flannel being installed along side Cilium. I also can't figure out if there's any way to remove kube-proxy's iptables after removing them. 

So far I have removed daemonsets, configmaps, and edited the machine config to not reference kube-proxy, as well as I have also patched it with cni set to none, so that cilium can be managed externally.

I'm also having a lot of unpredictable behavior having two CNI and the features duplicated between Cilium and kube-proxy/flannel.

Any help or ideas appreciated!

Thanks!

**Comment by smira** [comment, marked as answer]
Talos controlplane nodes (if you have more than one) try to sync bootstrap manifests at least on reboot, might do it on config change, etc.

So you want to disable Flannel and kube-proxy with the machine config.

You can verify that they are disabled with `talosctl get manifests -n <node>` for each controlplane node. They should all not have Flannel/kube-proxy manifests, after that things should be fine.

**Reply by jlowe64** [reply]
@smira I have my cni set to none in the machine config, however kube-flannel still installs itself. Would KubePrism have any dependencies causing this behavior?

**Reply by smira** [reply]
I described the steps above to see exactly which manifests Talos will try to install. Make sure all controlplane nodes are in agreement on that (no flannel manifest).

If you see the manifest for Flannel, `cni: ` is not set `none`.

**Reply by jlowe64** [reply]
@smira I think I managed to get the hang of it, the difference between the worker node and the control plane's machine configs were indeed the cni: none not being present on the worker node's running machine config.

**Reply by gecube** [reply]
Hi! We need a proper instructions with the precise steps how to remove flannel and kube-proxy on working cluster... Right now it is not clear how to properly remove them.

**Reply by smira** [reply]
Changing CNI on the fly is not something that I would recommend to do for anyone who doesn't feel to be a Kubernetes expert.

But the answer is above, there's nothing magic, it's all Kubernetes manifests that should be removed (from Kubernetes) after Talos part is disabled not to to try apply them back.

**Reply by gecube** [reply]
@smira Agree, but sometimes it is simpler than just to recreate cluster from the scratch. For me the most complex part is that the manifests for kube-proxy and flannel are hidden in configs and I need to guess how they were installed.

**Reply by smira** [reply]
They are not hidden, when enabled, `talosctl get manifests` shows them all in a pretty explicit way.

**Reply by gecube** [reply]
@smira but not after the changing the configuration... After changing machine description, the `talosctl get manifests` won't show them as it does not already manage them. Right?

**Reply by azkbn** [reply]
Got same issue, had to remove deamon sets manually
```
kubectl delete daemonset -n kube-system kube-flannel
kubectl delete daemonset -n kube-system kube-proxy
kubectl delete cm kube-flannel-cfg -n kube-system
```

**Reply by mulder999** [reply]
Same here, [documentation](https://www.talos.dev/v1.9/kubernetes-guides/network/deploying-cilium/) focuses on too many installation methods but completely overlooks the cleanup process.

**Reply by boldandbusted** [reply]
I've deleted any Flannel references in the Machineconfig. However, I get this:

```
$ talosctl get manifests
NODE            NAMESPACE      TYPE       ID                               VERSION
10.177.27.151   controlplane   Manifest   00-kubelet-bootstrapping-token   1
10.177.27.151   controlplane   Manifest   01-csr-approver-role-binding     1
10.177.27.151   controlplane   Manifest   01-csr-node-bootstrap            1
10.177.27.151   controlplane   Manifest   01-csr-renewal-role-binding      1
10.177.27.151   controlplane   Manifest   05-flannel                       1
[...]
```

How can I purge that `05-flannel` manifest? I can't edit it as `talosctl edit manifest 05-flannel` replies like so:

```
$ talosctl edit manifest 05-flannel
only the machineconfig resource can be edited
```

Some background: I'm trying to make sure that I can use `talosctl upgrade-k8s` without it trying to install kube-proxy again (I'm using Cilium). I'm purging all references to Flannel and kube-proxy, but upgrade-k8s still tries to upgrade kube-proxy - so I'm really trying to purge anything that might be causing these ghosts in my configuration.

Thanks.

**Reply by smira** [reply]
Lots of things changed, and Talos 1.13+ keeps the inventory, and it can do automatic cleanup with `talosctl upgrade-k8s`.

Any changes to Talos are done via machine configuration:

* https://docs.siderolabs.com/talos/v1.13/reference/configuration/v1alpha1/config#cni
* https://docs.siderolabs.com/talos/v1.13/reference/configuration/v1alpha1/config#proxy

```yaml
cluster:
  network:
    cni:
      name: none
  proxy:
    disabled: true
```
